src/internal/clusterstate/v2.10.0/pipelines_test.go

```diff
@@ -0,0 +1,175 @@
+				initial:       []version{1, 2, 3, 3},
+				hasDuplicates: true,
+			},
+		}
+	)
+	var db *pachsql.DB
+	c := pachd.NewTestPachd(t, pachd.TestPachdOption{
+		MutateFullOption: func(fullOption *pachd.FullOption) {
+			// run all but the last migration step
+			s := v2_10_0.MigratePartial(clusterstate.State_2_8_0, 1)
+			fullOption.DesiredState = &s
+		},
+		MutateEnv: func(env *pachd.Env) {
+			db = env.DB
+		},
+	})
+	for _, tc := range tcs {
+		t.Log("test case: ", tc.desc)
+		repo := "input"
+		pipeline := &pps.Pipeline{Name: "pipeline", Project: &pfs.Project{Name: pfs.DefaultProjectName}}
+		require.NoError(t, c.CreateRepo(pfs.DefaultProjectName, repo))
+		require.NoError(t, c.CreatePipeline(
+			pfs.DefaultProjectName,
+			"pipeline",
+			"", /* default image*/
+			[]string{"cp", "-r", "/pfs/in", "/pfs/out"},
+			nil, /* stdin */
+			nil, /* spec */
+			&pps.Input{Pfs: &pps.PFSInput{Project: pfs.DefaultProjectName, Repo: repo, Glob: "/*", Name: "in"}},
+			"",    /* output */
+			false, /* update */
+		))
+
+		// update the pipeline to create the necessary number of versions
+		for i := 0; i < len(tc.initial)-1; i++ {
+			_, err := c.RerunPipeline(c.Ctx(), &pps.RerunPipelineRequest{Pipeline: pipeline})
+			require.NoError(t, err)
+		}
+
+		// set database to bad state
+		tx := db.MustBeginTx(c.Ctx(), nil)
+		pKeys, pis := listPipelineVersions(t, c.Ctx(), tx, pipeline)
+		var pipUpdates []*v2_10_0.PipUpdateRow
+		pipVersionChanges := map[string]map[uint64]uint64{
+			pipeline.String(): {},
+		}
+		for i, v := range tc.initial {
+			if v != version(i)+1 {
+				pis[i].Version = uint64(v)
+				data, err := proto.Marshal(pis[i])
+				require.NoError(t, err)
+				vIdx := v2_10_0.VersionKey(pipeline.Project.Name, pipeline.Name, pis[i].Version)
+				update := &v2_10_0.PipUpdateRow{Key: pKeys[i], Proto: data, IdxVersion: vIdx}
+				pipUpdates = append(pipUpdates, update)
+				pipVersionChanges[pipeline.String()][uint64(i)] = uint64(v)
+			}
+		}
+		require.NoError(t, v2_10_0.UpdatePipelineRows(c.Ctx(), tx, pipUpdates))
+		require.NoError(t, v2_10_0.UpdateJobPipelineVersions(c.Ctx(), tx, pipVersionChanges))
+		require.NoError(t, tx.Commit())
+
+		// verify bad state
+		_, err := db.ExecContext(c.Ctx(), v2_10_0.CreateUniqueIndex)
+		if tc.hasDuplicates {
+			require.YesError(t, err)
+		} else {
+			require.NoError(t, err)
+			// now revert and continue test
+			_, err := db.Exec("DROP INDEX collections.pip_version_idx;")
+			require.NoError(t, err)
+		}
+
+		// run migration
+		tx = db.MustBeginTx(c.Ctx(), nil)
+		require.NoError(t, v2_10_0.DeduplicatePipelineVersions(c.Ctx(), migrations.Env{Tx: tx}))
+		require.NoError(t, tx.Commit())
+
+		// assert that there should be no duplicate pipelines detected in query
+		// assert that pipelines and jobs should list without error
+		pis, err = c.ListPipeline()
+		require.NoError(t, err)
+		require.Len(t, pis, 1)
+		jis, err := c.ListJob(pfs.DefaultProjectName, pipeline.Name, nil, -1, false)
```

Comment:
I think the `details=true` case is the one that breaks when there are duplicate pipelines.

Reply:
updating to `details=true`